R/container.R

```diff
@@ -0,0 +1,174 @@
+#' @param probabilities The names of class probability estimates (if any). For
+#' classification, these should be given in the order of the factor levels of
+#' the `estimate`.
+#' @param time The name of the predicted event time. (not yet supported)
+#' @examples
+#'
+#' container()
+#' @export
+container <- function(mode = "unknown", type = "unknown", outcome = character(0),
+                      estimate = character(0), probabilities = character(0),
+                      time = character(0)) {
+  dat <-
+    list(
+      outcome = outcome,
+      type = type,
+      estimate = estimate,
+      probabilities = probabilities,
+      time = time
+    )
+
+  new_container(
+    mode,
+    type,
+    operations = list(),
+    columns = dat,
+    ptype = tibble::new_tibble(list()),
+    call = current_env()
+  )
+}
+
+new_container <- function(mode, type, operations, columns, ptype, call) {
+  mode <- arg_match0(mode, c("unknown", "regression", "classification", "censored regression"))
+
+  if (mode == "regression") {
+    type <- "regression"
+  }
+
+  type <- arg_match0(type, c("unknown", "regression", "binary", "multiclass"))
+
+  if (!is.list(operations)) {
+    cli_abort("The {.arg operations} argument should be a list.", call = call)
+  }
+
+  is_oper <- purrr::map_lgl(operations, ~ inherits(.x, "operation"))
+  if (length(is_oper) > 0 & !any(is_oper)) {
+    bad_oper <- names(is_oper)[!is_oper]
+    cli_abort("The following {.arg operations} do not have the class \\
+                   {.val operation}: {bad_oper}.", call = call)
+  }
+
+  # validate operation order and check duplicates
+  validate_order(operations, mode, call)
+
+  # check columns
+  res <- list(
+    mode = mode, type = type, operations = operations,
+    columns = columns, ptype = ptype
+  )
+  class(res) <- "container"
+  res
+}
+
+#' @export
+print.container <- function(x, ...) {
+  # todo emulate Emil's recipe printing
+
+  num_op <- length(x$operations)
+  cli_inform("{x$type} post-processing object with {num_op} operation{?s}")
+
+  if (num_op > 0) {
+    cat("\n")
+    res <- purrr::map(x$operations, ~ print(.x))
+  }
+
+  invisible(x)
+}
+
+#' @export
+fit.container <- function(object, .data, outcome, estimate, probabilities = c(),
+                          time = c(), ...) {
+  # ------------------------------------------------------------------------------
+  # set columns via tidyselect
+
+  dat <- list()
+  dat$outcome <- names(tidyselect::eval_select(enquo(outcome), .data))
+  dat$estimate <- names(tidyselect::eval_select(enquo(estimate), .data))
+
+  probabilities <- tidyselect::eval_select(enquo(probabilities), .data)
+  if (length(probabilities) > 0) {
+    dat$probabilities <- names(probabilities)
+  } else {
+    dat$probabilities <- character(0)
+  }
+
+  time <- tidyselect::eval_select(enquo(time), .data)
+  if (length(time) > 0) {
+    dat$time <- names(time)
+  } else {
+    dat$time <- character(0)
+  }
```

Comment:
This would also allow for more easily implementing the "known selections" trick: https://github.com/tidymodels/yardstick/blob/be744a3d419398a165aa5bbea0e7f14e750d5b79/R/template.R#L832